openedx/core/djangoapps/courseware_api/tests/test_views.py

```diff
@@ -169,6 +181,37 @@ def test_course_metadata(self, logged_in, enrollment_mode, enable_anonymous, is_
             else:
                 assert not response.data['can_load_courseware']['has_access']
 
+    def test_happy_path_for_streak(self):
+        """ Test that metadata endpoint returns correct data for the streak celebration"""
+        CourseEnrollment.enroll(self.user, self.course.id, 'audit')
+
+        today = datetime.now(UTC)
+        for i in range(1, STREAK_LENGTH_TO_CELEBRATE + 1):
+            with mock.patch.object(UserCelebration, '_get_now') as get_now_mock:
+                get_now_mock.return_value = today + timedelta(days=i)
+                response = self.client.get(self.url, content_type='application/json')
+                should_celebrate = response.json()['celebrations']['should_celebrate_streak']
+                self.assertEqual(should_celebrate, i == STREAK_LENGTH_TO_CELEBRATE)
+
+    @mock.patch.object(UserCelebration, 'perform_streak_updates')
+    def test_streak_masquerade(self, perform_streak_updates_mock):
+        """ Don't update streak data when masquerading as a specific student """
+        self.user.is_staff = True
+        self.user.save()
+
+        user = UserFactory()
+        CourseEnrollment.enroll(user, self.course.id, 'verified')
+
+        self.update_masquerade(username=user.username)
+        now = datetime.now(UTC)
+        for i in range(1, STREAK_LENGTH_TO_CELEBRATE + 1):
+            with mock.patch.object(UserCelebration, '_get_now') as get_now_mock:
+                get_now_mock.return_value = now + timedelta(days=i)
+                response = self.client.get(self.url, content_type='application/json')
+                should_celebrate = response.json()['celebrations']['should_celebrate_streak']
+                self.assertEqual(should_celebrate, False)
+                perform_streak_updates_mock.assert_not_called()
```

Comment:
Yeah it would be. My approach there does duplicate some happy paths, at the cost of robustness against future changes. So it's a bit of a judgement call on whether it's worth it. But for cases like this, I think it's so easy to see how the tests might break in the future otherwise.

Let's say you start with:
```
def test_happy_path(self):
    setup_one_thing()
    assert result()

def test_one_disable_case(self):
    setup_one_thing()
    do_a_thing_that_actually_disables_it()
    assert not result()
```

Then in a month, another waffle gets added or some other gating logic like that. So now `test_happy_path` gets updated by an engineer that notices they broke `test_happy_path`:

```
def test_happy_path(self):
    setup_one_thing()
    setup_second_thing()
    assert result()
```

But they never update `test_one_disable_case` because it didn't break. And now the test isn't really doing what it intended to do, because the test setup isn't enabling the feature in the first place.

One way to avoid this is to do a pattern where setUp always does all setup the normal happy path needs. And then happy path tests are simply `assert result()`, leaving any disable- or error-paths to just do what they need on top. But in practice, things aren't always so clean. Especially with multiple happy paths.

This is all just a nit anyway. Just how I tend to write tests in large code bases where folks without context are likely to be coming in later and fixing broken tests, but not knowing what else might need to change.

In my approach, I'd write the test like so:
```
def test_one_disable_case(self):
    setup_one_thing()
    assert result()  # confirm we are working
    do_a_thing_that_actually_disables_it()
    assert not result()  # confirm we disabled it
```

Without that pattern, the test never really knows _why_ a thing is disabled, it just knows it _is_ disabled. Plus future engineers will be forced to notice if they break this case and fix the test.

Reply:
This might seem paranoid - I just have seen enough now-pointless, bitrotten tests in my life to want to be this cautious.

Reply:
sure that makes sense, i've updated the masquerade test with a true and then false case